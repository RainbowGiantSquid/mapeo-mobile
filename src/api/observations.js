// @flow
import { jsonRequest } from './base';
import { keyBy } from 'lodash';
import { parseObservationResponse } from '../models/observations';
import querystring from 'querystring'

import type {
  Observation as ObservationType,
  ServerObservationResponse,
  ServerObservationCreate
} from '../types/observation';

class Observation {
  static list = (opts) =>
    jsonRequest({
      method: 'GET',
      route: '/observations?' + querystring.stringify(opts)
    }).map(observations => {
      var payload = observations.map(parseObservationResponse)
      return keyBy(payload, 'id')
    });

  static create = (observation: ServerObservationCreate) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: observation
    }).map(parseObservationResponse);

  static update = (observation: ServerObservationResponse) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: observation
    }).map(parseObservationResponse);
}

export default Observation;
