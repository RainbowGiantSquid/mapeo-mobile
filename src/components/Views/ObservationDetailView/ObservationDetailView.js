// @flow
import React from 'react';
import { Button, Image, Text, TouchableHighlight, View, ScrollView, StyleSheet } from 'react-native';
import moment from 'moment';
import { NavigationActions, withNavigation } from 'react-navigation';
import MapboxGL from '@mapbox/react-native-mapbox-gl';
import type { Observation } from '@types/observation';

import LeftChevron from 'react-native-vector-icons/Entypo';
import ProfileImg from 'react-native-vector-icons/FontAwesome';

type State = {
  showStillHappening: boolean,
};

export type StateProps = {
  observations: {
    [id: string]: Observation,
  },
};

type Props = {
  navigation: NavigationActions;
};

const styles = StyleSheet.create({
  backChevron: {
    marginLeft: 15,
  },
  container: {
    backgroundColor: 'white',
    flex: 1,
    flexDirection: 'column',
  },
  date: {
    color: 'black',
    fontSize: 12,
    fontWeight: '600',
    textAlign: 'center',
  },
  header: {
    alignSelf: 'stretch',
    backgroundColor: '#ccffff',
    borderBottomColor: 'lightgray',
    borderBottomWidth: 1,
    height: 200,
    paddingVertical: 20,
  },
  mapBox: {
    flex: 1,
    alignSelf: 'stretch',
    marginBottom: 15,
    marginLeft: 15,
    marginRight: 15,
  },
  observedByText: {
    color: 'black',
    fontSize: 20,
    fontWeight: '700',
    paddingLeft: 10,
  },
  profileImage: {
    marginLeft: 10,
    marginBottom: 20,
  },
  section: {
    borderBottomColor: 'lightgray',
    borderBottomWidth: 1,
    flex: 1,
  },
  sectionText: {
    color: 'gray',
    fontSize: 12,
    marginBottom: 15,
    marginLeft: 15,
    marginTop: 15,
  },
  stillHappening: {
    color: 'black',
    fontSize: 12,
    fontWeight: '400',
    paddingTop: 15,
    paddingBottom: 15,
    textAlign: 'center',
  },
  textNotes: {
    color: 'black',
    fontSize: 18,
    fontWeight: '700',
    margin: 20,
    textAlign: 'center',
  },
  time: {
    color: 'grey',
    fontSize: 12,
    fontWeight: '300',
    textAlign: 'center',
  },
  title: {
    color: 'black',
    fontFamily: 'HelveticaNeue',
    fontSize: 18,
    fontWeight: '700',
    textAlign: 'center',
  },
});

class ObservationDetailView extends React.PureComponent<Props & StateProps, State> {
  state = { showStillHappening: true };

  dismissStillHappening = () => {
    if (this.state.showStillHappening) this.setState({ showStillHappening: false });
  }

  render() {
    const observation = this.props.observations[this.props.navigation.state.params.id];

    return (
      <ScrollView style={styles.container}>
        <View style={styles.header}>
          <TouchableHighlight onPress={() => { this.props.navigation.navigate('TabBarNavigation'); }}>
            <LeftChevron
              color="black"
              name="chevron-left"
              size={25}
              style={styles.backChevron}
            />
          </TouchableHighlight>
          <Text style={styles.title}>{observation.name}</Text>
          <Text style={styles.date}>{moment(observation.created).format('MMMM D, YYYY')}</Text>
          <Text style={styles.time}>{moment(observation.created).format('h:mm A')}</Text>
        </View>
        <View style={styles.section}>
          <Text style={styles.sectionText}>2 photos, 1 video</Text>
          <View style={{ flexDirection: 'row', height: 125 }}>
            <View style={{ flex: 1, backgroundColor: 'white' }} />
            <View style={{ flex: 1, backgroundColor: 'blue' }} />
            <View style={{ flex: 1, backgroundColor: 'lightgray' }} />
          </View>
          <Text style={styles.textNotes}>{observation.notes}</Text>
        </View>
        <View style={styles.section}>
          <Text style={styles.sectionText}>0.0 km</Text>
          <View style={{ height: 200 }}>
            <MapboxGL.MapView
              style={styles.mapBox}
              styleURL={MapboxGL.StyleURL.Street}
              zoomLevel={15}
              centerCoordinate={[11.256, 43.770]}
            />
          </View>
        </View>
        <View style={styles.section}>
          <Text style={styles.sectionText}>Observed by</Text>
          <View style={{ flexDirection: 'row' }}>
            <ProfileImg
              color="lightgray"
              name="user-circle-o"
              size={30}
              style={styles.profileImage}
            />
            <Text style={styles.observedByText}>{observation.observedBy}</Text>
          </View>
        </View>
        {
          this.state.showStillHappening ?
            <View style={{ flex: 1, backgroundColor: '#f2f2f2' }}>
              <Text style={styles.stillHappening}>Is this still happening here?</Text>
              <View style={{ flexDirection: 'row', justifyContent: 'center' }}>
                <Button onPress={this.dismissStillHappening} title="Yes" />
                <Button onPress={this.dismissStillHappening} title="No" />
              </View>
            </View> : null
        }
      </ScrollView>
    );
  }
}

export default withNavigation(ObservationDetailView);
