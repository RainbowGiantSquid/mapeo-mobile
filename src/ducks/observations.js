// @flow
import { create } from '@src/lib/redux';
import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { createSelector } from '@src/lib/selector';
import { StoreState } from '@types/redux';
import { Observation } from '@types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer,
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      observations: { $set: keyBy(action.payload, 'id') },
    });

    return newState;
  },
});

export const selectContact = createSelector(
  [
    (state: StoreState, id: string): Observation => state.app.observations[id],
  ],
  (observation: Observation): Observation => observation,
  (observation: Observation, id: string): string => id,
);

export default [observationListReducer];
