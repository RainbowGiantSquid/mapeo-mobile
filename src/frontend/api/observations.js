import fixture from "./test_observations.json";

export default class ObservationsApi {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
  }
  list(cb) {
    setTimeout(() => cb(null, fixture), 2000);
  }
  update() {}
  create() {}
}
