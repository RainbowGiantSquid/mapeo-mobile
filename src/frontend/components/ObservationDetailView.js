// @flow
import React from "react";
import { Text, Image, View, ScrollView, StyleSheet } from "react-native";

import FormattedCoords from "./FormattedCoords";
import ThumbnailScrollView from "./ThumbnailScrollVIew";
import InsetMapView from "./InsetMapView";
import ObservationIcon from "./ObservationIcon";
import LocationIcon from "./icons/LocationIcon";
// import UserIcon from "./icons/UserIcon";
// import CameraIcon from "./icons/CameraIcon";
import EditIcon from "./icons/EditIcon";
import type { PresetWithFields } from "../context/PresetsContext";
import type { Observation } from "../context/ObservationsContext";

const styles = StyleSheet.create({
  categoryIconContainer: {
    alignSelf: "center",
    position: "absolute",
    top: 25
  },
  categoryPin: {
    alignSelf: "center",
    width: 80,
    height: 85,
    marginTop: 5
  },
  container: {
    backgroundColor: "white",
    flex: 1,
    flexDirection: "column"
  },
  positionText: {
    fontSize: 12,
    color: "black",
    fontWeight: "700"
  },
  section: {
    borderBottomColor: "lightgray",
    borderBottomWidth: 1,
    flex: 1
  },
  sectionText: {
    color: "black",
    fontSize: 14,
    fontWeight: "700",
    marginBottom: 15,
    marginLeft: 10,
    marginTop: 15
  },
  textNotes: {
    color: "black",
    fontSize: 18,
    fontWeight: "700",
    margin: 20,
    textAlign: "center"
  },
  time: {
    color: "grey",
    fontSize: 12,
    fontWeight: "300",
    textAlign: "center"
  },
  fieldAnswer: {
    color: "black",
    fontSize: 18,
    fontWeight: "700",
    marginTop: 7,
    marginBottom: 15
  },
  fieldTitle: {
    color: "black",
    fontSize: 14,
    fontWeight: "700"
  }
});

const FieldView = ({ label, answer }) => (
  <View style={{ marginLeft: 15 }}>
    <Text style={styles.fieldTitle}>{label}</Text>
    <Text style={styles.fieldAnswer}>{answer || "Not entered"}</Text>
  </View>
);

type ODVProps = {
  observation: Observation,
  preset?: PresetWithFields
};

const ObservationDetailView = ({ observation, preset = {} }: ODVProps) => {
  const { lat, lon, attachments } = observation.value;
  const photos =
    attachments && attachments.filter(a => a.type === "image/jpeg");
  console.log(observation.value);
  return (
    <ScrollView style={styles.container}>
      <>
        <Image
          source={require("../images/category-pin.png")}
          style={styles.categoryPin}
        />
        <View style={styles.categoryIconContainer}>
          <ObservationIcon iconId={preset.icon} />
        </View>
        {lat != null && lon != null && (
          <View style={{ flexDirection: "row", alignSelf: "center" }}>
            <FormattedCoords lon={lon} lat={lat} style={styles.positionText} />
          </View>
        )}
        <Text style={styles.time}>{observation.createdAt}</Text>
        <View style={styles.section}>
          <Text style={styles.textNotes}>{observation.value.tags.notes}</Text>
        </View>
        {lat != null && lon != null && (
          <View style={styles.section}>
            <View
              style={{
                flexDirection: "row",
                alignItems: "center",
                marginLeft: 15
              }}
            >
              <LocationIcon size={20} />
              <FormattedCoords lon={lon} lat={lat} style={styles.sectionText} />
            </View>
            <InsetMapView style={{ height: 240 }} />
          </View>
        )}
        {/* $FlowFixMe - not sure why flow is not getting this type */}
        {photos && <ThumbnailScrollView photos={photos} />}
        {preset.fields && preset.fields.length && (
          <View style={styles.section}>
            <View
              style={{
                flexDirection: "row",
                alignItems: "center",
                marginLeft: 15
              }}
            >
              <EditIcon size={20} />
              <Text style={styles.sectionText}>Details</Text>
            </View>
            <>
              {preset.fields.map(({ label, key }) => (
                <FieldView
                  key={label}
                  label={label}
                  answer={observation.value.tags[key]}
                />
              ))}
            </>
          </View>
        )}
      </>
    </ScrollView>
  );
};

export default ObservationDetailView;
