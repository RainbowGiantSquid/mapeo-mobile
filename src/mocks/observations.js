// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 'ObservationID',
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: 'Thu Apr 05 2018 23:06:45 GMT+0200 (CEST)',
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: null,

  ...observation
});
