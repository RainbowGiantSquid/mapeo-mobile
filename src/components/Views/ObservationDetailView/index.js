// @flow
import { connect } from 'react-redux';
import type { Dispatch } from 'redux';
import { NavigationActions } from 'react-navigation';
import { StoreState } from '../../../types/redux';

import { observationAdd } from '../../../ducks/observations';
import ObservationDetailView from './ObservationDetailView';
import type { StateProps, DispatchProps } from './ObservationDetailView';

const resetAction = NavigationActions.navigate({
  routeName: 'TabBarNavigation',
  params: {},
  actions: [
    NavigationActions.navigate({
      routeName: 'CameraView',
      params: { showCancelButton: false }
    })
  ]
});

function mapStateToProps(state: StoreState): StateProps {
  return {
    observations: state.app.observations,
    selectedObservation: state.app.selectedObservation
  };
}

function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
  return {
    addObservation: observation => dispatch(observationAdd(observation)),
    goToPhotoView: source =>
      dispatch(
        NavigationActions.navigate({
          routeName: 'PhotoView',
          params: { photoSource: source }
        })
      ),
    resetNavigation: () => dispatch(resetAction)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(
  ObservationDetailView
);
