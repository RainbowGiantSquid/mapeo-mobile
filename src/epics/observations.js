// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import { OBSERVATION_LIST, observationList } from '../ducks/observations';
import { Action } from '../types/redux';
import { Observation } from '../types/observation';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, Observation[]>>,
  store: any
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(
      action =>
        action.status === 'Start' &&
        store.getState().app &&
        store.getState().app.observations &&
        !store.getState().app.observations.length
    )
    .map(() =>
      observationList('', [
        {
          type: 'Old houses and farms',
          id: 1,
          lat: 0.106831547,
          lon: -77.42499473,
          link: 'link',
          created: new Date('2018-03-25T12:00:00Z'),
          name: 'Old houses and farms',
          notes: `Pedro and Maria's house`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/1-house.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Hunting sites',
          id: 2,
          lat: 0.088170932,
          lon: -77.44020885,
          link: 'link',
          created: new Date('2018-03-25T12:30:00Z'),
          name: 'Hunting sites',
          notes: `Don Pancho's hunting camp`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/4-camp.jpg')
            }
          ],
          icon: require('../images/categories/category_2.png'),
          mock: true
        },
        {
          type: 'Threat',
          id: 3,
          lat: 0.112411067,
          lon: -77.41378756,
          link: 'link',
          created: new Date('2018-02-03T17:30:00Z'),
          name: 'Threat',
          notes: `Oil spill 3 February 2018. Ruptured pipeline. Dead fish, oily water`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/5-threat.jpg')
            }
          ],
          icon: null,
          mock: true
        },
        {
          type: 'Medicinal plants',
          id: 4,
          lat: 0.09749025,
          lon: -77.43614404,
          link: 'link',
          created: new Date('2018-02-04T17:30:00Z'),
          name: 'Medicinal plants',
          notes: `Salola fruit plant - used in tea to heal inflamation`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/6-plant.jpg')
            }
          ],
          icon: require('../images/categories/category_6.png'),
          mock: true
        },
        {
          type: 'Waterfalls',
          id: 5,
          lat: 0.105818148,
          lon: -77.4408037,
          link: 'link',
          created: new Date('2018-03-04T17:30:00Z'),
          name: 'Waterfalls',
          notes: `Sacred waterfall, visited during vision quests for ancestors, for taking ayahuasca`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/7-waterfall.jpg')
            }
          ],
          icon: require('../images/categories/category_13.png'),
          mock: true
        },
        {
          type: 'Rivers and streams',
          id: 6,
          lat: 0.119004333,
          lon: -77.42374225,
          link: 'link',
          created: new Date('2018-03-24T17:30:00Z'),
          name: 'Rivers and streams',
          notes: `Point where pipeline crosses river`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/8-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Historical sites',
          id: 7,
          lat: 0.082856811,
          lon: -77.4429836,
          link: 'link',
          created: new Date('2018-03-24T12:30:00Z'),
          name: 'Historical sites',
          notes: `There used to be a village here in the 1940s, it was abandoned after an epidemic of flu which killed most of the elders and children.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/9-specialsite.jpg')
            }
          ],
          icon: require('../images/categories/category_17.png'),
          mock: true
        },
        {
          type: 'Gold',
          id: 8,
          lat: 0.108605553,
          lon: -77.41560494,
          link: 'link',
          created: new Date('2018-03-11T17:30:00Z'),
          name: 'Gold',
          notes: `Illegal gold miners 11 March 2018, excavators, dredges, open tailing pond, mercury.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/10-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_8.png'),
          mock: true
        }
      ])
    );

export default [observationListEpic];
