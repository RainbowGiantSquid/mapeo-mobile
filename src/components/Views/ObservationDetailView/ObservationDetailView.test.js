// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import ObservationDetailView from './ObservationDetailView';
import { createObservation } from '../../../mocks/observations';

describe.only('ObservationDetailView tests', () => {
  test('snapshots', () => {
    const observation = createObservation();
    const isFocused = () => true;
    const addListener = () => true;

    let tree;
    tree = shallow(
      <ObservationDetailView
        navigation={{ isFocused, addListener }}
        selectedObservation={observation}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
