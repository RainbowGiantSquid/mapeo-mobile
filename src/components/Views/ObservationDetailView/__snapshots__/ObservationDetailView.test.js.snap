// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservationDetailView tests snapshots 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withNavigation(withNavigationFocus(ObservationDetailView))
    navigation={
      Object {
        "addListener": [Function],
        "isFocused": [Function],
      }
    }
    selectedObservation={
      Object {
        "categoryId": 19,
        "created": "2018-03-26T02:30:00.000Z",
        "icon": Object {
          "testUri": "../../../src/images/categories/category_18.png",
        },
        "id": 1,
        "lat": 50.5,
        "link": "linkID",
        "lon": 50.5,
        "media": Array [],
        "name": "Oil Spill",
        "notes": "This is pretty bad. It smells pretty noxious, and the entire pond is affected.",
        "observedBy": "You",
        "type": "Observation",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
